Sources/ComposedUI/CollectionView/CollectionCoordinator.swift

```diff
@@ -229,88 +245,289 @@ extension CollectionCoordinator: SectionProviderMappingDelegate {
+                guard batchRemovedIndexPath.section == removedIndexPath.section else { return batchRemovedIndexPath }
+
+                if batchRemovedIndexPath.item > removedIndexPath.item {
+                    return IndexPath(item: batchRemovedIndexPath.item - 1, section: batchRemovedIndexPath.section)
+                } else {
+                    return batchRemovedIndexPath
+                }
+            }
+        }
+
+        batchedRowRemovals.append(contentsOf: indexPaths)
+
+        for removedIndexPath in indexPaths {
+            batchedRowUpdates = batchedRowUpdates.map { updatedIndexPath in
+                guard updatedIndexPath.section == removedIndexPath.section else { return updatedIndexPath }
+
+                if updatedIndexPath.item > removedIndexPath.item {
+                    return IndexPath(item: updatedIndexPath.item - 1, section: updatedIndexPath.section)
+                } else {
+                    return updatedIndexPath
+                }
+            }
+
+            batchedRowInserts = batchedRowInserts.map { insertedIndexPath in
+                guard insertedIndexPath.section == removedIndexPath.section else { return insertedIndexPath }
+
+                if insertedIndexPath.item > removedIndexPath.item {
+                    return IndexPath(item: insertedIndexPath.item - 1, section: insertedIndexPath.section)
+                } else {
+                    return insertedIndexPath
+                }
+            }
+
+            batchedRowMoves = batchedRowMoves.map { element in
+                var (from, to) = element
+
+                if from.section == removedIndexPath.section, from.item > removedIndexPath.item {
+                    from.item -= 1
+                }
+
+                if to.section == removedIndexPath.section, to.item > removedIndexPath.item {
+                    to.item -= 1
+                }
+
+                return (from, to)
+            }
         }
-        if defersUpdate { return }
-        mappingDidEndUpdating(mapping)
     }
 
     public func mapping(_ mapping: SectionProviderMapping, didUpdateElementsAt indexPaths: [IndexPath]) {
         assert(Thread.isMainThread)
-        changes.append { [weak self] in
-            guard let self = self else { return }
-            
+
+        guard isPerformingBatchedUpdates else {
+            prepareSections()
+
             var indexPathsToReload: [IndexPath] = []
             for indexPath in indexPaths {
                 guard let section = self.sectionProvider.sections[indexPath.section] as? CollectionUpdateHandler,
-                    !section.prefersReload(forElementAt: indexPath.item),
-                    let cell = self.collectionView.cellForItem(at: indexPath) else {
-                        indexPathsToReload.append(indexPath)
-                        continue
+                      !section.prefersReload(forElementAt: indexPath.item),
```

Comment:
?

Reply:
This is how Xcode indents when copy/pasting or using re-indent.

I've reverted it, although I think it's clearer when some extra line breaks are added and running re-indent:

```swift
guard
    let section = self.sectionProvider.sections[indexPath.section] as? CollectionUpdateHandler,
    !section.prefersReload(forElementAt: indexPath.item),
    let cell = self.collectionView.cellForItem(at: indexPath)
else {
    indexPathsToReload.append(indexPath)
    continue
}
```

Reply:
I think that's fine.